baseapp/baseapp_test.go

```diff
@@ -823,3 +845,53 @@ func TestTxGasLimits(t *testing.T) {
 		}
 	}
 }
+
+func TestBaseAppAnteHandler(t *testing.T) {
+	anteKey := []byte("ante-key")
+	anteOpt := func(bapp *BaseApp) {
+		bapp.SetAnteHandler(anteHandlerTxTest(t, capKey1, anteKey))
+	}
+
+	deliverKey := []byte("deliver-key")
+	routerOpt := func(bapp *BaseApp) {
+		bapp.Router().AddRoute(routeMsgCounter, handlerMsgCounter(t, capKey1, deliverKey))
+	}
+
+	cdc := codec.New()
+	app := setupBaseApp(t, anteOpt, routerOpt)
+
+	app.InitChain(abci.RequestInitChain{})
+	registerTestCodec(cdc)
+	app.BeginBlock(abci.RequestBeginBlock{})
+
+	// execute a tx that will fail ante handler execution
+	//
+	// NOTE: No state should be persisted here which will be implicitly checked by
+	// the next test.
+	tx := newTxCounter(0, 0)
+	tx.setAnteFail(true)
+	txBytes, err := cdc.MarshalBinaryLengthPrefixed(tx)
+	require.NoError(t, err)
+	res := app.DeliverTx(txBytes)
+	require.False(t, res.IsOK(), fmt.Sprintf("%v", res))
+
+	// execute at tx that will fail message execution (state should mutate)
+	tx = newTxCounter(0, 0)
+	tx.setMsgFail(true)
+	txBytes, err = cdc.MarshalBinaryLengthPrefixed(tx)
+	require.NoError(t, err)
+	res = app.DeliverTx(txBytes)
+	require.False(t, res.IsOK(), fmt.Sprintf("%v", res))
+
+	// execute a successful ante handler and message execution where state is
+	// implicitly checked by previous tx executions
+	tx = newTxCounter(1, 0)
```

Comment:
Ok so I guess the first arg, 1, means that the ante handler for one of the two above succeeded?  But it's not clear which one.

The second 0 means that none of the msgs succeeded, but that wasn't clear either...

Reply:
I've added explicit store checks via getting the app's state. Lmk if this still isn't clear. This logic is baked into the testing ante handler and message handler, so I'm trying to work with what I already have without making further drastic changes.